### DISCUSSION TITLE
Is it safe to have branches from team members checked out

### DISCUSSION THREAD

**Discussion by uloco** [gitext-rs/git-stack#230] [Q&A] [answered]
It happened to me yesterday, that I had a branch a colleague was working on checked out locally. Not aware of this I ran `git stack --pull` and it rebased also the branch which does not belong to me.

What will happen if I `git stack --push` now?

Do I have to cleanup every branches of other people I occasionally checkout so I don't mess up their branches?

I assume `--force-with-lease` is used internally for `git stack --push`, but what if they haven't pushed anything yet but added commits locally?
Will `git-stack --pull` resolve the diverging histories for them? Or do they have to manually check out the remote branch (with a different name) and rebase their local changes here on top of it again?

What do you do in those scenarios?

**Comment by epage** [comment, marked as answer]
git-stack identifies and protects non-personal branches with the following caveats
- The branch contains a commit with you as the committer or author
- You've made a branch on top of the other user's branch
  - This seems like a bug, we should allow you to make branches on top of other people's work
- HEAD is on a commit on that branch

Overall, this was trying to have git-stack ignore other people's branches while allowing you to take over the branch and start doing work on it.

**Reply by uloco** [reply]
I did not create a branch on top of my team members one. I just checked their branch out and ran `git stack --pull`. This rebased their branch on top of the correct protected one as it would for a branch I was working on myself.

**Reply by uloco** [reply]
I think I had checked it out and HEAD was pointing to it yes. But this is really unfortunate...
Is this something you could fix in general or is this by design?

**Reply by epage** [reply]
Some different inputs into this
- Implementation: We have the same logic for marking branches as protected as abbreviating them.  We don't want to abbreviate the branch you are on.  This is easy to adjust.
- Like I had said, we are trying to balance preventing accidental modifications while allowing people to take them over.   With us no longer checking HEAD, for a user to rebase, they'd either need to make a commit first, use another tool like `git rebase`. or we'd need to add a flag to override it which might be hard to communicate and the more knobs we offer, the harder it is for users to find what they need.

**Reply by uloco** [reply]
I am not sure how your workflow is but taking over a branch is something we never do. And if I need to really do this, I would look into the docs to see what to do. I actually first looked what git-stack does here or if there is any comment about branches not belonging to me but couldn't find anything about this anyway.

But we do check out other peoples' branches all the time...
I am not sure if I can get my team to adapt this tool if they need to know details like this.
But if there is no better way for now, I'll try to get used to checking out protected branches before I do git stack --pull.

**Reply by uloco** [reply]
Thanks anyway!

**Reply by epage** [reply]
To be clear,. that isn't a "no" but a "this needs some things figured out".

I'd recommend opening an issue for further exploration of this.

**Reply by epage** [reply]
As for people's workflows, I've more so taken over people' s branches as an open source maintainer rather than someone working on a team as the expectations for contributors tends to be lower.

Even on a team, it happens from time to time whether due to collaborating, handing off work, etc.

**Reply by epage** [reply]
(I'll also add that I'm someone who never checks out other's people's work as part of the code review though I know there are people who do it for every review)

**Reply by uloco** [reply]
ok thanks for your in detail thoughts about this. will open an issue for this and we can discuss :)
